File: codes/golang/57.插入区间.go


func insert(intervals [][]int, newInterval []int) [][]int {
	startInt := newInterval[0]
	endInt := newInterval[1]
	var result [][]int
	placed := false
	for _, interval := range intervals {
		if interval[0] > endInt {
			if !placed {
				result = append(result, []int{startInt, endInt})
				placed = true
			}
			result = append(result, interval)
		} else if interval[1] < startInt {
			result = append(result, interval)
		} else {
			startInt = min(startInt, interval[0])
			endInt = max(endInt, interval[1])
		}
	}
	if !placed {
		result = append(result, []int{startInt, endInt})
	}
	return result
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

File: codes/java/57.插入区间.java

import java.util.ArrayList;
import java.util.List;

class Solution {
	public int[][] insert(int[][] intervals, int[] newInterval) {
		int start = newInterval[0];
		int end = newInterval[1];
		List<int[]> res = new ArrayList<>();
		boolean placed = false;
		for (int[] interval : intervals) {
			if (interval[0] > end) {
				if (!placed) {
					res.add(new int[] { start, end });
					placed = true;
				}
				res.add(interval);
			} else if (interval[1] < start) {
				res.add(interval);
			} else {
				start = Math.min(start, interval[0]);
				end = Math.max(end, interval[1]);
			}
		}
		if (!placed) {
			res.add(new int[] { start, end });
		}
		return res.toArray(new int[0][0]);
	}
}
